rust/datafusion/src/physical_plan/regex_expressions.rs

```diff
@@ -0,0 +1,161 @@
+            .zip(pattern_array.iter())
+            .zip(replacement_array.iter())
+            .map(|((string, pattern), replacement)| match (string, pattern, replacement) {
+                (None, _, _) => Ok(None),
+                (_, None, _) => Ok(None),
+                (_, _, None) => Ok(None),
+                (Some(string), Some(pattern), Some(replacement)) => {
+                    let replacement = regex_replace_posix_groups(replacement);
+
+                    // if patterns hashmap already has regexp then use else else create and return
+                    let re = match patterns.get(pattern) {
+                        Some(re) => Ok(re.clone()),
+                        None => {
+                            match Regex::new(pattern) {
+                                Ok(re) => {
+                                    patterns.insert(pattern.to_string(), re.clone());
+                                    Ok(re)
+                                },
+                                Err(err) => Err(DataFusionError::Execution(err.to_string())),
+                            }
+                        }
+                    };
+
+                    Some(re.map(|re| re.replace(string, replacement.as_str()))).transpose()
+                }
+            })
+            .collect::<Result<GenericStringArray<T>>>()?;
+
+            Ok(Arc::new(result) as ArrayRef)
+        }
+        4 => {
+            let string_array = downcast_string_arg!(args[0], "string", T);
+            let pattern_array = downcast_string_arg!(args[1], "pattern", T);
+            let replacement_array = downcast_string_arg!(args[2], "replacement", T);
+            let flags_array = downcast_string_arg!(args[3], "flags", T);
+
+            let result = string_array
+            .iter()
+            .zip(pattern_array.iter())
+            .zip(replacement_array.iter())
+            .zip(flags_array.iter())
+            .map(|(((string, pattern), replacement), flags)| match (string, pattern, replacement, flags) {
+                (None, _, _, _) => Ok(None),
+                (_, None, _, _) => Ok(None),
+                (_, _, None, _) => Ok(None),
+                (_, _, _, None) => Ok(None),
+                (Some(string), Some(pattern), Some(replacement), Some(flags)) => {
+                    let replacement = regex_replace_posix_groups(replacement);
+
+                    // format flags into rust pattern
+                    let (pattern, replace_all) = if flags == "g" {
+                        (pattern.to_string(), true)
+                    } else if flags.contains('g') {
+                        (format!("(?{}){}", flags.to_string().replace("g", ""), pattern), true)
+                    } else {
+                        (format!("(?{}){}", flags, pattern), false)
+                    };
+
+                    // if patterns hashmap already has regexp then use else else create and return
+                    let re = match patterns.get(&pattern) {
```

Comment:
I was worried that the key into the hashmap might not include the flags, but that seems not the be the case 

This clause seems basically copy/paste from the `3` case above -- maybe (in a future PR) or something we could consolidate them into a helper or something

Reply:
The code at line 115 does overload `pattern` (the key) with the flags. There is going to be opportunity to make some helper functions to manage these flags as we add more functions.